src/actions/groups.cr

```diff
@@ -1,34 +1,171 @@
 module DppmRestApi::Actions::Groups
+  class NoSuchGroup < NotFound
+    def initialize(context : HTTP::Server::Context, id : String, cause : Exception? = nil)
+      super context, "No group found with ID " + id, cause
+    end
+  end
+
+  class InvalidAccessLevel < BadRequest
+    def initialize(context : HTTP::Server::Context, cause : Exception? = nil)
+      level = context.params.url["access_level"]? || context.params.query["access_level"]?
+      super context, %[Given access level "#{level}" is not valid], cause
+    end
+  end
+
+  def query_params_in_body?(of context : HTTP::Server::Context) : Hash(String, Array(String))
+    if body = context.request.body
+      Hash(String, Array(String)).from_json body
+    else
+      {} of String => Array(String)
+    end
+  rescue e : JSON::ParseException
+    raise BadRequest.new context, <<-HERE, cause: e
+      Expected either a valid JSON Hash of String to Array of String, or no
+      specified request body at #{context.request.path}
+      HERE
+  end
+
   extend self
   include RouteHelpers
   relative_post do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Create
-    # TODO: create a group based on the response body
-
+    body = context.request.body
+    raise BadRequest.new context, "One must specify a group to add." if body.nil?
+    group = Config::Group.from_json body
+    DppmRestApi.permissions_config.groups << group
+    DppmRestApi.permissions_config.sync_to_disk
+    build_json context.response do |response|
+      response.field("successfullyAddedGroup") { group.to_json response }
+    end
   end
-  relative_put "/:id/route" do |context|
+  # Regardless of whether the given path exists on the given group, set the
+  # group's access level on that path to the given access level.
+  relative_put "/:id/route/:path/:access_level" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Update
-    # TODO: add a new route to the group
-
+    group_id_s = context.params.url["id"]
+    group_id = group_id_s.to_i? || raise BadRequest.new context,
+      %[Group ID #{group_id_s} was not a valid integer]
+    # find the group if it already exists in the permissions config
+    index = DppmRestApi.permissions_config
+      .groups
+      .index { |grp| grp.id == group_id }
+    raise NoSuchGroup.new context, group_id_s if index.nil?
```

Comment:
`find` would avoid to have to do a second lookup.
The above logic can also be in a method, that can be reused the other `delete` endpoints too.

Reply:
I attempted the refactor to do this last night, and it didn't work. The problem is that since all of our data-types are structs, changing a found group doesn't update that group in the config, only the local copy of it. We need to find the index, get the value at the index, then reinsert the modified value back in... or switch to classes.

Thoughts?

Reply:
I can't be a struct issue, nothing is passed around, and most importantly Array is always a reference.
Here the deal is because it's an array, that's why an index is needed.

Reply:
When you do...
```
group = DppmRestApi.permissions_config
      .groups
      .find { |grp| grp.id == group_id } || raise "nil guard"
group.change_something
```
...then...
```crystal
DppmRestApi.permissions_config
      .groups
      .find { |grp| grp.id == group_id }.should_not eq group
```
...whereas if group were a class, they would still be equivalent. Does that make sense? Or am I mistaken?

Reply:
`Group#permissions` is a Hash, so a reference.
Try to use this, this will return a class instead of a copied struct:
```cr
DppmRestApi.permissions_config
      .groups
      .find { |grp| grp.id == group_id }
      .try &.permissions
```

Reply:
`Group#permissions` is a Hash, so a reference.
Try to use this, this will return a class instead of a copied struct:
```cr
DppmRestApi.permissions_config
      .groups
      .find { |grp| grp.id == group_id }
      .try &.permissions
```

Reply:
And what bout changing the name? The ID of a group should be immutable, but the name not.
Or, we can make `Group` totally immutable (no property), and we have to pass through `DppmRestApi::Config` to create one, that will check if the ID and the name already exist.